### PR TITLE
#485 Make ImagePullPolicy configurable

### LINKED ISSUE #485: default ImagePullPolicy setting for keycloak manifest (Keycloak Operator)
### Describe the bug

Hey All,
I am using keycloak operator 16.1.0 and it comes with imagepullpolicy as always for keycloak [statefullset ](https://github.com/keycloak/keycloak-operator/blob/main/deploy/crds/keycloak.org_keycloaks_crd.yaml)(keycloak.yaml).
Also it is not defined in the CRD of the keycloak how to specify image pull policy.
Currently, in my case, the keycloak image is present in the private registry and I am pulling from there but, as the imagepullpolicy set as always it is pulling the image every time even if the image is present locally on the node.

Is my understanding correct about this policy. If yes, Can anyone help out if I can modify the policy?

Thank you.

### Version

16.1.0

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Anything else?
Below is my keycloak manifest.

```
apiVersion: keycloak.org/v1alpha1
kind: Keycloak
metadata:
  labels:
      app: test-keycloak
  name: test-keycloak
  namespace: test
spec:
  extensions:
    - ...
  externalDatabase:
    enabled: true
  instances: 1
  externalAccess:
    enabled: false
  keycloakDeploymentSpec:
    experimental:
      env:
        ...
```

### PR DESCRIPTION
**Pull request by IoannisMatzaris**
## Related GH Issue
Closes #485 

## Additional Information
- InitContainers & Containers of keycloak & postgres can have an ImagePullPolicy as user input in the CRD
- Unit tests updated
- pre-commit hook updated go.mod & go.sum
- Example in deploy/examples/keycloak/keycloak.yaml

## Verification Steps

1. make setup -> make code/gen -> kubectl apply -f deploy/examples/keycloak/keycloak.yaml
2. Check the imagePullPolicy in the keycloak statefulSet

## Checklist:
- [x] Automated Tests
- [ ] [Documentation](https://github.com/keycloak/keycloak-documentation) changes if necessary 
- [ ] (If required) Discussed on [Keycloak DEV](https://groups.google.com/forum/#!forum/keycloak-dev)

### CODE REVIEW

File: pkg/model/keycloak_deployment.go [hunk 1]
```diff
@@ -246,13 +246,15 @@ func KeycloakDeployment(cr *v1alpha1.Keycloak, dbSecret *v1.Secret, dbSSLSecret
...
 							},
-							VolumeMounts:   KeycloakVolumeMounts(cr, KeycloakExtensionPath, dbSSLSecret, KeycloakCertificatePath),
-							LivenessProbe:  livenessProbe(),
-							ReadinessProbe: readinessProbe(),
-							Env:            getKeycloakEnv(cr, dbSecret),
-							Args:           cr.Spec.KeycloakDeploymentSpec.Experimental.Args,
-							Command:        cr.Spec.KeycloakDeploymentSpec.Experimental.Command,
-							Resources:      getResources(cr),
+
+							ImagePullPolicy: cr.Spec.KeycloakDeploymentSpec.ImagePullPolicy,
```
Comment (vmuzikar) [inline pkg/model/keycloak_deployment.go:250]: Please add the pull policy also to `KeycloakDeploymentReconciled`.

File: tools.go [hunk 2]
```diff
@@ -1,3 +1,4 @@
+//go:build tools
```
Comment (vmuzikar) [inline tools.go:1]: Can you please elaborate why to make this change?

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@IoannisMatzaris Thank you for your contribution! Added a few  comments.

File: tools.go (hunk 2 above)
Comment (IoannisMatzaris) [reply tools.go:1]: This comes from go 1.17.6 & `make code/fix`(_goimports_). 

[More info](https://go.dev/doc/go1.17#go-command): -> Search for "_//go:build lines_"

According to the docs, this does not seem to introduce any breaking changes with older versions.

File: go.mod [hunk 3]
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/sirupsen/logrus v1.6.0
 	github.com/spf13/pflag v1.0.5
 	github.com/stretchr/testify v1.6.1
+	golang.org/x/tools v0.1.10 // indirect
```
[the remark this answers is not in the document]
Comment (IoannisMatzaris) [reply go.mod]: The dependency was removed

File: pkg/model/keycloak_deployment_test.go [hunk 4]
```diff
@@ -123,6 +127,23 @@ func testExperimentalArgs(t *testing.T, deploymentFunction createDeploymentState
 	assert.Equal(t, []string{"test"}, args)
 }
 
+func testDeploymentSpecImagePolicy(t *testing.T, deploymentFunction createDeploymentStatefulSet) {
```
[the remark this answers is not in the document]
Comment (IoannisMatzaris) [reply pkg/model/keycloak_deployment_test.go:130]: good idea! done :+1:

File: pkg/model/keycloak_deployment_test.go [hunk 5]
```diff
@@ -123,6 +127,23 @@ func testExperimentalArgs(t *testing.T, deploymentFunction createDeploymentState
...
+	//given
+	dbSecret := &v1.Secret{}
+	cr := &v1alpha1.Keycloak{}
+	cr.Spec.KeycloakDeploymentSpec = v1alpha1.KeycloakDeploymentSpec{
+		DeploymentSpec: v1alpha1.DeploymentSpec{
+			ImagePullPolicy: v1.PullNever,
```
[the remark this answers is not in the document]
Comment (IoannisMatzaris) [reply pkg/model/keycloak_deployment_test.go:136]: I added an e2e test on this [commit](https://github.com/keycloak/keycloak-operator/pull/506/commits/eb32cf62038c55ec090867205774e84121e1e122)

**Comment by IoannisMatzaris** [comment]
@vmuzikar Thank you for your feedback! 

I added some more commits. Could you please take a look?

**Review by vmuzikar — CHANGES_REQUESTED** [review]
@IoannisMatzaris Thanks for the update.

Added two (hopefully last :)) comments. Apart from that, can you please rebase and fix the linter errors?

File: go.sum [hunk 6]
```diff
@@ -1692,7 +1692,6 @@ k8s.io/gengo v0.0.0-20190128074634-0689ccc1d7d6/go.mod h1:ezvh/TsK7cY6rbqRK0oQQ8
 k8s.io/gengo v0.0.0-20190822140433-26a664648505/go.mod h1:ezvh/TsK7cY6rbqRK0oQQ8IAqLxYwwyPxAX1Pzy0ii0=
 k8s.io/gengo v0.0.0-20200114144118-36b2048a9120/go.mod h1:ezvh/TsK7cY6rbqRK0oQQ8IAqLxYwwyPxAX1Pzy0ii0=
 k8s.io/gengo v0.0.0-20200413195148-3a45101e95ac/go.mod h1:ezvh/TsK7cY6rbqRK0oQQ8IAqLxYwwyPxAX1Pzy0ii0=
-k8s.io/gengo v0.0.0-20201113003025-83324d819ded h1:JApXBKYyB7l9xx+DK7/+mFjC7A9Bt5A93FPvFD0HIFE=
```
[the remark this answers is not in the document]
Comment (IoannisMatzaris) [reply go.sum:1695]: This change comes from `go mod tidy`. I would suggest to keep it.

Should I reverse it?

File: test/e2e/keycloaks_test.go [hunk 7]
```diff
@@ -67,6 +67,17 @@ func NewKeycloaksSSLTestStruct() *CRDTestStruct {
 	}
 }
 
+func NewKeycloaksWithDefaultImagePullPolicyTestStruct() *CRDTestStruct {
+	return &CRDTestStruct{
+		prepareEnvironmentSteps: []environmentInitializationStep{
+			prepareKeycloaksCR,
+		},
+		testSteps: map[string]deployedOperatorTestStep{
+			"keycloakWithPodLabelsDeploymentTest": {testFunction: keycloakDeploymentDefaultImagePullPolicyTest},
```
[the remark this answers is not in the document]
Comment (IoannisMatzaris) [reply test/e2e/keycloaks_test.go]: 07b3a13dbc638115969f968884b385d06b9fb0d4

**Comment by IoannisMatzaris** [comment]
@vmuzikar Thank you very much for your feedback :smiley: [One last thing](https://github.com/keycloak/keycloak-operator/pull/506#discussion_r855737938) before rebasing!

**Review by vmuzikar — APPROVED** [review]
@IoannisMatzaris LGTM, thanks!

**Pull request merged**